fastchat/train/train_lora.py

```diff
@@ -145,10 +145,22 @@ def train():
         trainer.train()
     trainer.save_state()
 
-    # Save states. Weights might be a placeholder in zero3 and need a gather
-    state_dict = get_peft_state_maybe_zero_3(
+     # check if zero3 mode enabled
+    if trainer.hf_deepspeed_config_orig.is_zero3():
+        # use deepspeed engine internal function to gather state dict
+        # state_dict_zero3 contains whole parameters of base and lora adapters
+        # we will not extract lora parameters since peft save_pretrained will do that
+        # https://github.com/huggingface/peft/blob/3714aa2fff158fdfa637b2b65952580801d890b2/src/peft/peft_model.py#L125
+        # https://github.com/huggingface/peft/blob/3714aa2fff158fdfa637b2b65952580801d890b2/src/peft/utils/save_and_load.py#L19
+        state_dict_zero3 = trainer.model_wrapped._zero3_consolidated_16bit_state_dict()
```

Comment:
Seems like this one gathers the weight of the whole model, while we'd hope only to gather that of the LoRA(For instance, when training a 30B model with 2x40GB A100, gathering the whole 30B model needs at least 60GB memory, resulting in an OOM).

Reply:
Hi, 
thanks for the rely!
@ZYHowell 

according to deepspeed explanation:["works on one layer at a time to require as little gpu0 memory as possible, by moving the already consolidated weights to cpu" ](https://github.com/microsoft/DeepSpeed/blob/e02b8d0bd06fd5e3d65c4bc410708ee2443334c7/deepspeed/runtime/engine.py#L3244-L3254)


It will not gathering all parameters to gpu0 memory at the same time and will not cause OOM, since it work layer by layer and will move a gathered layer to cpu before start to gather another layer.

Reply:
On the CPU there is another concern: deepspeed launches a process for each GPU, and I wonder if this function gathers a weight on each process, or just the rank 0 process. If it's the first case, the total CPU memory consumption will be #parameter(60~350GB) x #GPU(4 or 8), which is even a burden for servers with large memory.

Reply:
Hi,

only rank 0 have a full state_dict and other rank just get a None. 

    def _zero3_consolidated_16bit_state_dict(self):
        """
        Get a full non-partitioned state_dict with fp16 weights on cpu.
        Important: this function must be called on all ranks and not just rank 0.
        This is similar to nn.Module.state_dict (modelled after _save_to_state_dict), but:
        1. consolidates the weights from different partitions on gpu0
        2. works on one layer at a time to require as little gpu0 memory as possible, by
        moving the already consolidated weights to cpu
        3. takes care to keep the shared params shared when gradually copying the params to cpu
        Returns:
            **a consolidated fp16 ``state_dict`` on cpu on rank 0, ``None`` on other ranks**
        """